### PR TITLE
Motion benchmarking system

### PR DESCRIPTION
**Pull request by liquidmetal9015**
# Draft

## Description
This is a suite of tools to perform benchmarks of robot motion and performance. To give quantifiable feedback on how well the robots are performing in as many dimensions as possible in a procedural environment.

So, there are a lot of concerning things about the way this runs right now and it needs work to really be good. 

Some issues:
One of the functions requires numpy
All end of motion metrics are broken due to the way move terminates and when they are recorded
The way some functions are called is a bit jank
Needs a whole lot of documentation all over the place
The non-basic motion tests aren't written yet
It isn't really doing anything with the outputted data
Imports are a bit jank due to some issues I was having
Generally many of the metrics are borked due to various little issues

## Associated Issue
Issue #1318

## Design Documents
[Link](https://docs.google.com/document/d/1uA66JDNMuNfRuFQLymauNZkwIWz0kD3KOiUOSZp76vE/edit?usp=sharing)

### Test Case 1
1. Look at the motion_benchmark play to figure out what you want to run
2. run RunMotionBenchmark in soccer

Expected result: You get some kind of useful information?

### CODE REVIEW

**Comment by kylestach** [comment]
@liquidmetal9015 is this worth reviewing in its current state?

File: soccer/gameplay/plays/testing/latency_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import skills.move
+import skills.capture
+import tactics.coordinated_pass
+import constants
+import main
+import enum
+import tools
+import skills.latency_test
+
+class RunLatencyTest(play.Play):
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/latency_test.py:13]: comments about what the point of the file is

File: soccer/gameplay/plays/testing/latency_test.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import robocup
+import play
+import behavior
+import skills.move
+import skills.capture
+import tactics.coordinated_pass
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/latency_test.py:6]: Separate imports like [here](https://github.com/RoboJackets/robocup-software/blob/staging/soccer/gameplay/plays/offense/adaptive_formation.py#L5)

File: soccer/gameplay/plays/testing/motion_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import robocup
+import play
+import behavior
+import skills.move
+import skills.capture
+import tactics.coordinated_pass
+import constants
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/motion_benchmark.py:7]: separate imports like [here](https://github.com/RoboJackets/robocup-software/blob/staging/soccer/gameplay/plays/offense/adaptive_formation.py#L5)

File: soccer/gameplay/plays/testing/motion_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import main
+import enum
+import tools
+import skills.latency_test
+import skills.benchmark.motion_benchmark
+import skills.benchmark.basic_motion_test
+import robocup
+
+class RunMotionBenchmark(play.Play):
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/motion_benchmark.py:16]: Comment what's the point of the file

File: soccer/gameplay/plays/testing/motion_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        midFace2.facePoints.append(robocup.Point(0,2.8))
+        midFace2.facePoints.append(robocup.Point(0,0))
+        midFace2.facePoints.append(robocup.Point(2,0))
+        midFace2.title = "Mid size triangle while facing points #2"
+        self.benchmark.addTest(midFace2)
+
+
+        #Pure rotational Tests do not currently work
+        '''
+        pureRot = skills.benchmark.basic_motion_test.BasicMotionTest(numberOfRuns, self.benchmark)
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/motion_benchmark.py:122]: Remove the comment and make an issue.

File: soccer/gameplay/skills/benchmark/ball_proximity_test.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#Test of pivoting around the ball
+#May use Line kick or move, I'm not sure yet
+class BallProximityTest(MotionTest):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/ball_proximity_test.py:3]: I dont agree with adding nonfunctional code. I think you should create an issue

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
+import skills.benchmark.motion_test
+import time
+import skills.benchmark.motion_benchmark
+#import Motion
+#import motion_test.MotionTest
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:5]: Delete comments + extra linds

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+
+
+
+##Runs motion tests involving polygonal motions with or without a face command
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:10]: do you mean with and without? meaning that it runs both?

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+        self.posEndError = [0.0] * self.motions
+        self.lineFollowError = [0.0] * self.motions
+        self.rotationalFollowError = [0.0] * self.motions
+        self.finalRotationalError = [0.0] * self.motions
+        self.maxOvershoot = [0.0] * self.motions
+        self.endVel = [0.0] * self.motions
+        self.totalVel = [0.0] * self.motions
+        self.motionNumber = -1
+        #print("This test has: " + str(self.runs) + " runs")
+        #print("This test has: " + str(self.positions) + " positions")
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:101]: extra (?) prints. Please remove if you don't want them

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    # Appends a point to the list of face points
+    def addFacePoint(self, point):
+        self.facePoints.append(a)
+
+
+    #Sets up the test to be run
+    #Largly initalizes arrays to arrays of zeros
+    def setupTest(self):
+        print("Setting up test: " + self.title)
+        #if(len(self.points) == 0 or len(self.facePoints) == 1 or len(self.points) == len(self.facePoints)):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:87]: remove comments

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+                self.currentFacePoint = None
+            self.startIndex += 1
+            self.endIndex += 1
+            self.faceIndex += 1
+
+        self.motionStartTime = time.time()
+        self.timeOfLastCalc = time.time()
+
+
+    def updateDistances():
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:151]: comments for method description

File: soccer/gameplay/skills/benchmark/basic_motion_test.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+    
+    #A function that processes the results and stores them to the relavant maps
+    #This function causes a seemingly totally unrelated error
+    def processResults(self):
+        print("We are actually processing results")
+        return  
+        timeTaken = self.timeTaken
+        #avgMotionTime = sum(self.timeTaken) / len(self.timeTaken)
+        #motionTimeVar = statistics.pvariance(self.timeTaken)
+        #avgEndPosError = sum(self.posEndError) / len(self.posEndError)
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_motion_test.py:254]: Delete comments

File: soccer/gameplay/skills/benchmark/basic_triangle_motion_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+
+
+class BasicTriangleMotionBenchmark(BaseMotionBenchmark):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/basic_triangle_motion_benchmark.py:3]: What does this class do? lmao does it do anything?

File: soccer/gameplay/skills/benchmark/latency_test.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
+import single_robot_composite_behavior
+import behavior
+from enum import Enum
+import main
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/latency_test.py:4]: please import like [here](https://github.com/RoboJackets/robocup-software/blob/staging/soccer/gameplay/plays/offense/adaptive_formation.py#L5)

File: soccer/gameplay/skills/benchmark/latency_test.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        self.add_subbehavior(skills.move.Move(move_point), 'move') 
+
+    def on_exit_setup(self):
+        self.remove_all_subbehaviors()
+
+    def on_enter_noise(self):
+        self.noiseStartTime = time.time()
+        self.noiseStartPos = self.robot.pos
+
+    def execute_noise(self):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/latency_test.py:93]: Please describe method in comment

File: soccer/gameplay/skills/benchmark/latency_test.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        if(deltaX > self.noiseMaxX):
+            self.noiseMaxX = deltaX
+        if(deltaX < self.noiseMinX):
+            self.noiseMinX = deltaX
+        if(deltaY < self.noiseMinY):
+            self.noiseMinY = deltaY
+        if(deltaY > self.noiseMaxY):
+            self.noiseMaxY = deltaY
+
+    def brokenNoise(self):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/latency_test.py:107]: Please describe method in comment

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
+import single_robot_composite_behavior
+import single_robot_behavior
+import behavior
+from enum import Enum
+import main
+import constants
+import robocup
+import skills.move
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:8]: imports

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+    ##
+    # This function used to do something
+    #
+    def setupTests(self):
+       pass 
+
+    def getVel(self):
+        return self.robot.vel
+
+    def getSpeed(self):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:161]: lmao i feel like these functions (getSpeed, getAngleError, getPosError) should be implemented in c++ for every robot. correct me if im wrong though

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+        return math.sqrt(xErr**2 + yErr**2)
+
+    def getLineError(self,start, end):
+        startNumpy=np.array([start.x,start.y])
+        endNumpy=np.array([end.x,end.y])
+        robotNumpy=np.array([self.robot.pos.x,self.robot.pos.y])
+        d = np.cross(startNumpy-endNumpy,endNumpy-robotNumpy)/np.linalg.norm(endNumpy-startNumpy)
+        return abs(d)
+ 
+    def getOvershoot(self, start, end):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:183]: Describe method in comment

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+        #pass
+
+
+    #End setup state functions
+
+    #Noise state functions
+
+   #def on_enter_noise(self):
+   #     self.noiseStartTime = time.time()
+   #     self.noiseStartPos = self.robot.pos
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:217]: delete comments

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+    def on_exit_TestEnd(self):
+        self.testIndex += 1
+        if(self.testIndex < len(self.tests)):
+            self.currentTest = self.tests[self.testIndex]
+            #self.currentTest.setupTest()
+        else:
+            self.currentTest = None
+
+    #We need to
+    def on_enter_ProcessAllTests(self):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:306]: I don't know what you intended to do here, but delete extra comments after you do

File: soccer/gameplay/skills/benchmark/motion_benchmark.py [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+                print("Its a dict")
+
+        return c
+
+
+        #This is what was in the role requiremnts before, what I have now works but is not ideal
+        
+        '''
+        a = role_assignment.RoleRequirements()
+        print("--------------------------ROLE_REQUIREMENTS_THING--------------------")
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_benchmark.py:356]: delete extra comments

File: soccer/gameplay/skills/benchmark/motion_test.py [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+        #The time at which the last iterative calculation was made
+        self.timeOfLastCalc = 0.0
+
+        #The number of test runs remaining
+        #self.runs = 0
+
+
+    #Just a distance calculation
+    def calcDist(self, point1, point2):
```
Comment (Jason27chan) [inline soccer/gameplay/skills/benchmark/motion_test.py:40]: function for distance between robocup points [here](https://github.com/RoboJackets/robocup-software/blob/staging/soccer/gameplay/robocup-py.cpp#L710)

**Review by Jason27chan — CHANGES_REQUESTED** [review]
lmao

**Comment by liquidmetal9015** [comment]
I think I'm going to rework this a bit and make some smaller PRs in the future.

**Pull request closed without merging**